src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs

```diff
@@ -523,13 +523,57 @@ protected virtual void Dispose(bool disposing)
                 }
 
                 _server?.Dispose();
-                _host?.StopAsync().Wait();
+                _host?.StopAsync().GetAwaiter().GetResult();
                 _host?.Dispose();
             }
 
             _disposed = true;
         }
 
+        /// <inheritdoc />
+        public async ValueTask DisposeAsync()
+        {
+            await DisposeAsyncCore();
+
+            Dispose(disposing: false);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources asynchronously
+        /// </summary>
+        protected virtual async ValueTask DisposeAsyncCore()
```

Comment:
We shouldn't need to add this API. I think it's sufficient to explicitly implement `IAsyncDisposable` in derived class and call `base.`

Reply:
You mean, I should remove this method?

Reply:
Yes

Reply:
But then dispose will be synchronous! Because public method `DisposeAsync()` will call sync version of dispose on line 538.

Reply:
I think Dispose should call the DisposeAsync and do GetAwaiter().GetResult() instead and make this method private instead.

Reply:
@pranavkm I have removed `DispostAsyncCore()` method in the latest commit. But right now `DisposeAsync()` method calling the synchronous version of `Dispose()` under the hood, isn't that wrong? Or am I missing something here to understand! 
Thanks!